### PR TITLE
Feat/aiagents4pharma docker ci/cd

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

### for talk2aiagents4pharma
image

### for talk2knowledgegraphs
image

## Description
This PR introduces a **Docker Compose setup** and a **Dockerfile** to containerize and manage the `Talk2AIAgents4Pharma`  and `Talk2Knowledgegraphs` application/agent.  

-> Also added CI/CD for `Talk2AIAgents4Pharma` and `Talk2Knowledgegraphs` 
- Direct build and push to docker hub similar to pr #118
- Action were tested on my fork and application was tested locally

Docker Compose configuration that sets up and connects
- Ollama with nomic-embed-text model

### Docker Compose Configuration
- Created a `docker-compose.yml` that defines both services in an isolated network
- Set up proper service dependencies to ensure correct startup order
- Configured port mappings (11434 for Ollama API, 8501 for Streamlit app)
- Implemented persistent volume storage for Ollama models

### Ollama Service
- Uses official `ollama/ollama:latest` image
- Automatically pulls the `nomic-embed-text` embedding model at startup
- Includes health check to verify service availability
- Uses shell-based entrypoint for running multiple startup commands

### Talk2AIAgents4Pharma Service(similar for Talk2Knowledgegraphs)
- Uses the pre-built image `virtualpatientengine/talk2aiagents4pharma:latest` 
- Connects to Ollama service through the internal Docker network
- Loads environment variables from `.env` file for API keys

### Environment Setup
- Added `.env` template for required credentials:
  - OpenAI API key
  - NVIDIA API key
  - Ollama host configuration

## Testing
The setup has been tested to verify:
- Both services start correctly
- Network connectivity between services works
- Ollama successfully loads the embedding model
- Streamlit application is accessible via browser

## How to Use
1. Clone the repository
2. Add your API keys to the `.env` file
3. Run `docker-compose up -d` (# in the specific path i.e: aiagents4pharma/talk2aigents4pharma)
4. Access the Streamlit app at http://localhost:8501

#### **For Users**
```sh
docker-compose up -d
```
Check logs with:  
```sh
docker-compose logs -f
```
Verify Ollama-nomic is running:  
```sh
curl -X POST http://localhost:11434/api/embeddings -d '{
  "model": "nomic-embed-text",
  "prompt": "This is a final test for embedding confirmation."
}' | jq
```

Output(This is just for logging and testing purposes): 
```sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 15247    0 15150  100    97  41308    264 --:--:-- --:--:-- --:--:-- 41658
{
  "embedding": [
    1.2225806713104248,
    0.5954437255859375,
    .....
    .....
    -2.458794355392456,
    -1.5912129878997803,
    -0.7651652097702026,
    -0.5293824672698975,
    -0.2852611243724823,
    -0.46075016260147095
  ]
}
```

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Review by gurdeep330 — APPROVED** [review]
I can see that someone has a strong liking for containerization. 😊

Great effort! 💪

File: README.md [hunk 1]
```diff
@@ -37,19 +37,64 @@ Check out the tutorials on each agent for detailed instrcutions.
...
+
+2. Copy the `.env.example` file and rename it to `.env`:
+   ```sh
+   cp .env.example .env
+   ```
+3. Open the `.env` file and add your API keys:
+   ```plaintext
+   OPENAI_API_KEY=your_openai_api_key
+   NVIDIA_API_KEY=your_nvidia_api_key
+   # Note: You only need to add OPENAI_API_KEY_API_KEY for Talk2KnowledgeGraphs
```
Comment (awmulyadi) [inline README.md:64]: There are possibly typos in the scripts and comments: `talk2aiagents4harma` and  `OPENAI_API_KEY_API_KEY`. Please kindly re-check each command line so that new potential contributors are able to run the script without hassle 👍  A follow-up question is that whether talk2aiagents4pharma doesn't need this `OPENAI_API_KEY`?

File: docs/index.md [hunk 2]
```diff
@@ -37,19 +37,64 @@ Check out the tutorials on each agent for detailed instrcutions.
...
+
+2. Copy the `.env.example` file and rename it to `.env`:
+   ```sh
+   cp .env.example .env
+   ```
+3. Open the `.env` file and add your API keys:
+   ```plaintext
+   OPENAI_API_KEY=your_openai_api_key
+   NVIDIA_API_KEY=your_nvidia_api_key
+   # Note: You only need to add OPENAI_API_KEY_API_KEY for Talk2KnowledgeGraphs
```
Comment (awmulyadi) [inline docs/index.md:64]: Make sure to update this file as well 👍

**Review by awmulyadi — CHANGES_REQUESTED** [review]
Hi, @ansh-info! Great work for this contenarization PR 🥇 . I just have few minor comments regarding typos.

File: README.md (hunk 1 above)
Comment (ansh-info) [reply README.md:64]: Thank you for pointing that out! 😊 *talk2aiagents4pharma* requires an `OPEN_API_KEY`. Regarding the note, it was written that way because the key is only needed if the user wants to use *talk2knowledgegraphs* but not *talk2aiagents4pharma*.

File: docs/index.md (hunk 2 above)
Comment (ansh-info) [reply docs/index.md:64]: Sure

File: README.md (hunk 1 above)
Comment (ansh-info) [reply README.md:64]: If have update the note like this, please let me know its fine:
   ```plaintext
   # Note: You need both of these keys for Talk2AIAgents4Pharma
   OPENAI_API_KEY=your_openai_api_key
   NVIDIA_API_KEY=your_nvidia_api_key
   # Note: The API end point for ollama is already added in the env.example
   OLLAMA_HOST=http://ollama:11434
   # Note: You only need to add OPENAI_API_KEY for Talk2KnowledgeGraphs if running it separately
   ```

Comment (awmulyadi) [reply README.md]: The comments are more clear now 👍

**Review by awmulyadi — APPROVED** [review]
Looks good to me 👍

File: aiagents4pharma/talk2aiagents4pharma/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    # Add healthcheck to ensure the service is running properly
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:11434/api/version"]
+      interval: 30s
+      timeout: 10s
+      retries: 3
+    # Start the Ollama server
+    entrypoint: ["/bin/sh", "-c"]
+    command: >
+      "ollama serve &
```
[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2aiagents4pharma/docker-compose.yml:19]: Your observation is correct. By default, the command runs the CPU version of Ollama. To enable GPU support, we need to add specific options:

- **For NVIDIA GPUs:**  
  We include a section (using Docker Compose's deploy resources) that reserves all GPUs (equivalent to using the `--gpus=all` flag in a `docker run` command).

- **For AMD GPUs:**  
  We map the required devices (`/dev/kfd` and `/dev/dri`) and use the `ollama/ollama:rocm` image as recommended.

In our updated setup, we define separate services (or profiles) for CPU, NVIDIA, and AMD variants. This means you choose which variant to run by activating the corresponding profile (e.g., `--profile nvidia` for NVIDIA GPUs). If no profile is specified, the CPU version runs by default.

This unified configuration allows you to deploy the appropriate version without modifying environment variables, and ensures GPU passthrough is correctly set up when needed.


- To start the containers, run the following command from the root directory:

```sh
docker compose --profile cpu up # for CPU mode
docker compose --profile nvidia up # for GPU mode
docker compose --profile amd up # for AMD mode
```

[the remark this answers is not in the document]
Comment (ansh-info) [reply aiagents4pharma/talk2aiagents4pharma/docker-compose.yml:19]: @dmccloskey Right now, the CPU version is working perfectly. @gurdeep330 and I are actively working on enabling NVIDIA support. We ran the curl command and confirmed that Ollama is set up correctly, but we need a little more time to debug the NVIDIA configuration. I have already opened an issue regarding this, and I have created a new issue #129 and subsequent pull request once we resolve it.

**Review by dmccloskey — APPROVED** [review]
